crates/dbsp/src/storage/buffer_cache/cache.rs

```diff
@@ -229,78 +226,176 @@ where
+            return Ok(aux.clone());
+        }
+
+        let (sender, receiver) = oneshot::channel();
+
+        {
+            let mut requests = self.requests.lock().unwrap();
+            let (size2, senders) = requests
+                .entry(location.offset)
+                .or_insert((location.size, Vec::new()));
+            debug_assert_eq!(location.size, *size2);
+            senders.push(sender);
+        }
+
+        self.n_requests.send_modify(|n| *n += 1);
+        let block = receiver.await.unwrap()?; // XXX unwrap
+        let aux = parse(block, location)?;
+        self.cache.inner.lock().unwrap().insert(key, aux.clone());
+        Ok(aux)
     }
-    fn create_named(&self, name: &Path) -> Result<Box<dyn FileWriter>, StorageError> {
-        Self::backend().create_named(name)
+
+    /// Waits until `goal` threads have blocked on I/O in [Self::read].
+    pub async fn wait(&self, goal: usize) {
+        self.n_requests
+            .subscribe()
+            .wait_for(|n| *n >= goal)
+            .await
+            .unwrap();
     }
 
-    fn open(&self, name: &Path) -> Result<Arc<dyn FileReader>, StorageError> {
-        Self::backend().open(name)
+    /// Runs all of the pending I/O and wakes up threads blocked in [Self::read].
+    pub async fn run_io_batch<R>(&self, file: &R)
+    where
+        R: FileReader + ?Sized,
+    {
+        let requests = std::mem::take(&mut *self.requests.lock().unwrap());
+        let n_requests = requests
+            .values()
+            .map(|(_offset, result_senders)| result_senders.len())
+            .sum::<usize>();
+        self.n_requests.send_modify(|n| *n -= n_requests);
+        let blocks = requests
+            .iter()
+            .map(|(offset, (size, _))| *offset..*offset + *size as u64)
+            .collect::<Vec<_>>();
+        let (sender, receiver) = oneshot::channel();
+        file.read_async(
+            blocks,
+            Box::new(
+                move |result| sender.send(result).unwrap(), // XXX unwrap
+            ),
+        );
+        let result = receiver.await.unwrap(); // XXX unwrap
+        for (result, (_, result_senders)) in result.into_iter().zip(requests.into_values()) {
+            for result_sender in result_senders {
+                result_sender.send(result.clone()).unwrap(); // XXX unwrap
+            }
+        }
     }
 
-    fn base(&self) -> PathBuf {
-        Self::backend().base()
+    /// Execute all of the `tasks` on `file` until all of them run to
+    /// completion, returning a vector of their return values in the same order.
+    ///
+    /// Internally, this runs in a series of rounds, where in each round we run
+    /// each task until it either completes or blocks on I/O on `file`. At the
+    /// end of the round, if any tasks are still left, we do all of the I/O on
+    /// all of the tasks in a single batch of reads.
+    pub async fn execute_tasks<F, T, R>(
```

Comment:
it seems odd that we have this function to await things when await is already a lang mechanism (just an observation)